File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        HashSet<Character> seen = new HashSet();
        int l=0, r=0;
        int max =0;

        while(r<s.length()){
            char c = s.charAt(r);
            if(seen.add(c)){
                max= Math.max(max,r-l+1);
                r++;
            }else{
               while (s.charAt(l)!=c){
                  seen.remove(s.charAt(l));
                  l++;
               }
               seen.remove(c);
               l++;
            }
        }
        return max;
        
    }
}

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:

       ans = 0
       sett = set()
       l = 0

       for i in range(len(s)):
        while s[i] in sett:
            sett.remove(s[l])
            l+=1
        sett.add(s[i])
        ans = max(ans,i-l +1)
        
       return ans
            





       
       
       
       
       
       
       
       
       
       
       
       
       
       
       
       
       
       
       
       
       
       
       
       
       
       
       
       
       
       
       
       
       
       
       
       
       
       
       
       
       
       
       
       
       
       
        # charSet = set()
        # l = 0
        # ans = 0

        # for r in range(len(s)):
        #     while s[r] in charSet:
        #         charSet.remove(s[l])
        #         l+=1
        #     charSet.add(s[r])
        #     ans = max(ans,r-l+1)
        # return ans
            
        